src/internals/ir/callGraph.ts

```diff
@@ -0,0 +1,227 @@
+        this.processStatements(func.statements, func.id);
+      }
+    }
+  }
+
+  private getFunctionName(
+    func: AstFunctionDef | AstReceiver | AstContractInit,
+  ): string | undefined {
+    switch (func.kind) {
+      case "function_def":
+        return func.name.text;
+      case "contract_init":
+        return `contract_init_${func.id}`;
+      case "receiver":
+        return `receiver_${func.id}`;
+      default:
+        unreachable(func);
+    }
+  }
+
+  private processStatements(statements: AstStatement[], callerId: number) {
+    for (const stmt of statements) {
+      switch (stmt.kind) {
+        case "statement_expression":
+          this.processExpression(stmt.expression, callerId);
+          break;
+        case "statement_condition":
+          this.processStatements(stmt.trueStatements, callerId);
+          if (stmt.falseStatements) {
+            this.processStatements(stmt.falseStatements, callerId);
+          }
+          break;
+        case "statement_while":
+        case "statement_until":
+        case "statement_repeat":
+        case "statement_foreach":
+          this.processStatements(stmt.statements, callerId);
+          break;
+        case "statement_try":
+          this.processStatements(stmt.statements, callerId);
+          break;
+        case "statement_try_catch":
+          this.processStatements(stmt.statements, callerId);
+          this.processStatements(stmt.catchStatements, callerId);
+          break;
+        case "statement_let":
+        case "statement_return":
+        case "statement_assign":
+        case "statement_augmentedassign":
+          if (stmt.expression) {
+            this.processExpression(stmt.expression, callerId);
+          }
+          break;
+        default:
+          console.warn(`Unhandled statement type: ${(stmt as any).kind}`, stmt);
+      }
+    }
+  }
+
+  private processExpression(
+    expr: AstExpression | AstStructFieldInitializer,
+    callerId: number,
+  ) {
+    this.forEachExpression(expr, (nestedExpr) => {
+      let calleeId: number | undefined;
+      if (nestedExpr.kind === "static_call") {
+        const staticCall = nestedExpr as AstStaticCall;
+        calleeId = this.findOrAddFunction(staticCall.function.text);
+      } else if (nestedExpr.kind === "method_call") {
+        const methodCall = nestedExpr as AstMethodCall;
+        calleeId = this.findOrAddFunction(methodCall.method.text);
+      }
+      if (calleeId !== undefined) {
+        this.addEdge(callerId, calleeId);
+      }
+    });
+  }
+
+  private forEachExpression(
+    expr: AstExpression | AstStructFieldInitializer,
+    callback: (expr: AstExpression | AstStructFieldInitializer) => void,
+  ) {
+    callback(expr);
+
+    if (typeof expr === "object" && expr !== null) {
+      if ("args" in expr && Array.isArray(expr.args)) {
+        for (const arg of expr.args) {
+          this.forEachExpression(arg, callback);
+        }
+      } else if ("value" in expr) {
```

Comment:
What does it mean? Why don't you make a type assertion here to get the concrete types of nodes supporting the `.value` field?

Reply:
It complicates the maintaining of this code.